valhalla/thor/pathalgorithm.h

```diff
@@ -159,5 +148,54 @@ struct EdgeMetadata {
   }
 };
 
+// A structure for tracking time information as the route progresses
+struct TimeInfo {
+  // whether or not the provided location had valid time information or not
+  bool valid;
+
+  // index into the timezone database of the location
+  // used to do timezone offset as the route progresses
+  int timezone_index;
+
+  // seconds from epoch adjusted for timezone at the location
+  // used to do local time offset as the route progresses
+  uint64_t local_time;
+
+  // the ordinal second from the beginning of the week (starting monday at 00:00)
+  // used to look up historical traffic as the route progresses
+  uint32_t second_of_week;
+
+  // the distance in seconds from now
+  int64_t seconds_from_now;
+
+  // helper function to initialize the object from a location
+  static TimeInfo make(valhalla::Location& location, baldr::GraphReader& reader);
+
+  // a tuple used to offset from a time in the forward or reverse direction
+  struct Offset {
+    float seconds;
+    int timezone_index;
+  };
+
+  // offset all the initial time info to reflect the progress along the route to this point
+  TimeInfo operator+(Offset offset) const;
+
+  // offset all the initial time info to reflect the progress along the route to this point
+  TimeInfo operator-(Offset offset) const;
+
+  // for unit tests
+  bool operator==(const TimeInfo& ti) const {
+    return valid == ti.valid && timezone_index == ti.timezone_index && local_time == ti.local_time &&
+           second_of_week == ti.second_of_week && seconds_from_now == ti.seconds_from_now;
+  }
+
+  // for unit tests
+  friend std::ostream& operator<<(std::ostream& os, const TimeInfo& ti) {
```

Comment:
i reckon this method only gets unit test coverage when the tests fail but it was super valuable for testing so i left it